ui/helpers/taskcluster.js

```diff
@@ -1,48 +1,123 @@
-import { OIDCCredentialAgent, Queue } from 'taskcluster-client-web';
+import { Queue } from 'taskcluster-client-web';
+import debounce from 'lodash/debounce';
+import delay from 'lodash/delay';
+import moment from 'moment';
 
-import { loginRootUrl, getUserSessionUrl } from './url';
+import { clientId, redirectURI } from '../taskcluster-auth-callback/constants';
+
+import { loginRootUrl, createQueryParams } from './url';
+
+export const tcCredentialsMessage =
+  'Need to retrieve or renew Taskcluster credentials before action can be performed.';
 
 const taskcluster = (() => {
-  let credentialAgent = null;
+  // TODO set default to firefox-ci rootUrl
+  let _rootUrl = 'https://hassan.taskcluster-dev.net/';
+
+  // from the MDN crypto.getRandomValues doc
+  const secureRandom = () =>
+    window.crypto.getRandomValues(new Uint32Array(1))[0] / 4294967295;
 
-  // Create an OIDC credential agent if it doesn't exist.
-  const tcAgent = () => {
-    if (credentialAgent) {
-      return credentialAgent;
+  const generateNonce = () => {
+    let value = '';
+    const characters =
+      'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
+
+    for (let i = 0; i <= 20; i++) {
+      value += characters.charAt(
+        Math.floor(secureRandom() * characters.length),
+      );
     }
 
-    const userSession = localStorage.getItem('userSession');
-    const oidcProvider = 'mozilla-auth0';
+    localStorage.setItem('requestState', value);
+    return value;
+  };
+
+  const getAuthCode = (useExistingWindow = false) => {
+    const nonce = generateNonce();
+    // we're storing these for use in the TaskclusterCallback component (taskcluster-auth.html)
+    // since that's the only way for it to get access to them
+    localStorage.setItem('requestState', nonce);
+    localStorage.setItem('tcRootUrl', _rootUrl);
+
+    // the expires param is optional and if it's greater than the maxExpires value that's been
+    // registered with the client (set to 3 days) or is omitted, expiry will default to maxExpires
+    const params = {
+      client_id: clientId,
+      response_type: 'code',
+      redirect_uri: redirectURI,
+      scope: 'treeherder',
+      state: nonce,
+    };
+
+    if (useExistingWindow) {
+      window.location.href = `${_rootUrl}login/oauth/authorize${createQueryParams(
+        params,
+      )}`;
+    } else {
+      window.open(
+        `${_rootUrl}login/oauth/authorize${createQueryParams(params)}`,
+        '_blank',
+      );
+    }
+  };
+
+  const getDebouncedAuthCode = debounce(getAuthCode, 500, {
+    leading: true,
+    trailing: false,
+  });
+
+  const getCredentials = rootUrl => {
+    const userCredentials = JSON.parse(localStorage.getItem('userCredentials'));
+    // TODO remove staging instance
+    _rootUrl =
+      rootUrl === loginRootUrl
+        ? 'https://hassan.taskcluster-dev.net/'
```

Comment:
It's possible that the root URL wouldn't work with a trailing slash so probably we can remove it. Same with the other places where the root url is set.

Reply:
I was wondering about that. We typically write urls with trailing stashes (assigned to const's) but maybe in this case I should leave that off and change the code accordingly.